src/dev/flang/ast/AbstractType.java

```diff
@@ -1562,6 +1564,64 @@ static boolean checkActualTypePars(SourcePosition pos, AbstractFeature called, L
     return result;
   }
 
+
+  /**
+   * collect all the directly and indirectly - via generics -
+   * used types in this type.
+   * example: Cons i32 (list i32) => { Cons, i32, list }
+   *
+   * @return
+   */
+  private Set<AbstractType> usedTypes()
+  {
+    var result = new TreeSet<AbstractType>();
+    usedTypes(result);
+    return result;
+  }
+
+
+  /**
+   * recursion helper for `usedTypes (no args)`
+   * @param result
+   */
+  private void usedTypes(TreeSet<AbstractType> result)
+  {
+    if (result.add(this))
+      {
+        for (AbstractType at : this.generics())
+          {
+            at.usedTypes(result);
+          }
+        if (isChoice())
+          {
+            for (AbstractType at : this.choiceGenerics())
+              {
+                at.usedTypes(result);
+              }
+          }
```

Comment:
I think this should not be needed, but it depends what you want. The current implementation for a type `t` in
```
  t : u a b is
```
woulbe be just `t` while with
```
  t : choice a b is
```
would be `t`, `a`, `b`.

I think you should only check the visibility of the direct type parameter, for inherited type parameters, this should be checked at the declaration of the type `t`: if `t` inherits from some `u` that is at least as visible as `t`, then all the type parameter of `u` should be as visible as `t`. So your do not need to check the inheritance at a use of `t`, also no need to check the choice generics.

Reply:
Maybe I misunderstand but this should be okay (`Sequence` being public):
```
private a : Sequence some_private_type is
```
while this is not
```
public a : Sequence some_private_type is
```
So I think it is in general not sufficient to just check direct type parameters?